projects/core/src/window/window-ref.ts

```diff
@@ -2,17 +2,41 @@ import { DOCUMENT } from '@angular/common';
 import { Inject, Injectable } from '@angular/core';
 import { fromEvent, Observable, of } from 'rxjs';
 import { debounceTime, distinctUntilChanged, startWith } from 'rxjs/operators';
-
+import { SERVER_REQUEST_ORIGIN, SERVER_REQUEST_URL } from '../util/ssr.tokens';
+/**
+ * Provides access to the window and document in a save way for both SSR and CRS.
+ *
+ * The document that is resolved during SSR will also be
+ */
 @Injectable({
   providedIn: 'root',
 })
 export class WindowRef {
   readonly document: Document;
 
-  constructor(@Inject(DOCUMENT) document) {
+  // TODO(#10467): make optional arguments mandatory with next major release
+  constructor(
+    @Inject(DOCUMENT) document,
+    @Inject(SERVER_REQUEST_URL) protected serverRequestUrl?: string,
+    @Inject(SERVER_REQUEST_ORIGIN) protected serverRequestOrigin?: string
+  ) {
     // it's a workaround to have document property properly typed
     // see: https://github.com/angular/angular/issues/15640
-    this.document = document;
+    this.document = this.resolveDocument(document);
+  }
+
+  /**
+   * Resolves the document and provides the href and origin in case they're undefined.
+   * This is useful in SSR, where those values are lacking.
+   */
+  protected resolveDocument(document: Document): Document {
+    if (!document.location) {
+      document.location = {
+        href: this.serverRequestUrl,
+        origin: this.serverRequestOrigin,
+      } as any;
```

Comment:
In SSR Angular Universal resolves document to a custom object implementation. I would not mutate it.

Anyway, SERVER_REQUEST_URL and _ORIGIN will need fixing, because of the recently discovered bug in ccv2 https://github.com/SAP/spartacus/issues/11016 . Looks like a bummer blocking this PR

Reply:
Blocker already fixed in PR https://github.com/SAP/spartacus/pull/11056